[WPE][GTK][WTF] Implement SystemTracing.h with Sysprof

GeorgesStavracas opened the pull request:
#### cb1a4a419f2189c993f8e1e8df93d08800750b0f
<pre>
[WPE][GTK][WTF] Implement SystemTracing.h with Sysprof
<a href="https://bugs.webkit.org/show_bug.cgi?id=275314">https://bugs.webkit.org/show_bug.cgi?id=275314</a>

Reviewed by Nikolas Zimmermann and Michael Catanzaro.

Implement the tracing macros provided by SystemTracing.h using the
Sysprof collector API.

Sysprof collector operates by passing a control fd as an environment
variable. This env var is set when running any executable with the
sysprof-cli utility, or through Sysprof&apos;s main interface.

The control fd needs to be propagated to child processes properly. For
the case of run-minibrowser script - a Python script that spawns the
actual MiniBrowser executable - and for the subprocess launcher. To do
that, duplicate the control fd and pass it as an environment variable
to subprocesses well. The Python script neatly supports that, but the
subprocess launcher needed some more manual mangling to get that done.

This initial work covers all pre-existing tracing groups implemented
on kdebug.

This patch does not introduce new tracing points.

There are limitations with the tracing macros though, for example,
they are most meaningful when there&apos;s a single WebKit view running.
Once more views are created, they all output into the same tracing
group. In the future, this can be fixed by using the subprocesses&apos;
pids as part of the tracing name, but right now it&apos;s most useless
as the sandbox makes all subprocesses have pid 3.

Another limitation is that tracing only truly works when disabling
the sandboxing of subprocesses. Working around it might require some
reengineering of the subprocess launching mechanism, so this is left
out of this patch as well.

To run WebKitGTK with Sysprof tracing, run the following command in
the WebKit source directory:

```
$ sysprof-cli --force -e WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 -- Tools/Scripts/run-minibrowser --gtk
```

For WPE WebKit, run the following command:

```
$ sysprof-cli --force -e WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 -- Tools/Scripts/run-minibrowser --wpe
```

It is also possible to get GNOME Shell marks by passing `--gnome-shell`
to the sysprof-cli command.

Co-authored-by: Carlos Garcia Campos &lt;cgarcia@igalia.com&gt;

* Source/JavaScriptCore/jit/JITPlan.cpp:
(JSC::JITPlan::compileInThread):
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/SystemTracing.h:
(WTF::tracePoint):
* Source/WTF/wtf/glib/SysprofAnnotator.h: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setReadyState):
* Source/WebKit/GPUProcess/glib/GPUProcessMainGLib.cpp:
(WebKit::GPUProcessMain):
* Source/WebKit/UIProcess/API/glib/WebKitInitialize.cpp:
(WebKit::webkitInitialize):
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/WebProcess/gtk/WebProcessMainGtk.cpp:
* Source/WebKit/WebProcess/wpe/WebProcessMainWPE.cpp:
* Tools/Scripts/webkitpy/common/system/abstractexecutive.py:
(AbstractExecutive.run_command):
* Tools/Scripts/webkitpy/common/system/executive.py:
(Executive.run_command):
* Tools/Scripts/webkitpy/common/system/executive_mock.py:
(MockExecutive.run_command):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.run_minibrowser):

Canonical link: <a href="https://commits.webkit.org/280690@main">https://commits.webkit.org/280690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/385fb1143e6fa9255d999e3bf9fd9e5461a3a900

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57334 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9809 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60956 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7777 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44286 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7967 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46428 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59364 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49514 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27291 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/56860 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6825 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6782 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50426 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62635 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56576 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1247 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7188 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53690 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49548 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53770 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12672 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1060 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78336 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32491 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12975 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33576 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34661 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33322 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+    }
+
+    SysprofAnnotator(SysprofAnnotator const&) = delete;
+    void operator=(SysprofAnnotator const&) = delete;
```
You could replace those with `WTF_MAKE_NONCOPYABLE()`, as in:

```suggestion
    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
```

aperezdc commented on Source/WebKit/GPUProcess/glib/GPUProcessMainGLib.cpp:
```diff
@@ -40,6 +44,9 @@ class GPUProcessMainGStreamer final: public AuxiliaryProcessMainBase<GPUProcess>
 {
+#if USE(SYSPROF_CAPTURE)
+    WTF::SysprofAnnotator::Initialize("WebKit (GPU)");
+#endif
```
This would be better inside the `ENABLE(GPU_PROCESS)` block right below.

aperezdc commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -211,6 +211,20 @@ void ProcessLauncher::launchProcess()
+    if (const char* sysprofFdStr = getenv("SYSPROF_CONTROL_FD"))
+        sysprofFd = atoi(sysprofFdStr);
+
+    if (sysprofFd > 0) {
```
While it's extremely unlikely to happen here, file descriptor zero is a valid one, so the check here should use `>= 0`:

```suggestion
    if (sysprofFd >= 0) {
```

aperezdc commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -211,6 +211,20 @@ void ProcessLauncher::launchProcess()
+    int sysprofFd = -1;
+
+    if (const char* sysprofFdStr = getenv("SYSPROF_CONTROL_FD"))
+        sysprofFd = atoi(sysprofFdStr);
```
We have a `parseInteger()` template in WTF that returns an optional, and that allows checking whether the input was valid, so I would do instead:

```suggestion
        sysprofFd = parseInteger<int>(sysprofFdStr).value_or(-1);
```

There's also `String::number(int)` to convert a number to a string but probably not worth using below as it would need the usual `.utf8().data()` dance to be passed to GLib functions.

aperezdc commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+    SysprofAnnotator() { };
+    ~SysprofAnnotator()
+    {
+        for (auto const & iter : ongoingMarks) {
```
I think the style checker will want you to remove the spaces at the left of the ampersand, and also we typically see `const auto` more often than `auto const` in WebKit code:

 ```suggestion
         for (const auto& iter : ongoingMarks) {
```

aperezdc commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+        if (instance.ongoingMarks.contains(key)) {
+            value = WTFMove(instance.ongoingMarks[key]);
+            instance.ongoingMarks.erase(key);
+        }
```
You can do the same in a more convenient way using `HashMap::take()`, which already removes the entry from the hash map and returns the value in one go:

```suggestion
        if (auto entry = instace.ongoingMarks.find(key))
            value = WTFMove(instance.ongoingMarks.take(entry));
```

GeorgesStavracas replied:
The thing here is that this is an `std::map`, not a `HashMap`. Mostly because I would have to write a hash function and I was lazy :P

aperezdc reviewed (changes requested):
Nice one!

We have an `WTF_ATTRIBUTE_PRINTF()` which should help placate the compiler warnings (being treated as errors, hence the EWS failure).

I left a few comments suggesting simplifications and alternative ways of doing things commonly used in WebKit; I hope you don't mind 😇

csaavedra commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+    }
+
+private:
+    static const char* tracePointCodeName(TracePointCode code)
```
This could use `ASCIILiteral` instead (recently we favor using `ASCIILiteral` instead of `const char*` whenever possible).

csaavedra commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+        }
+    }
+
+    std::string processName { "WebKit" };
```
shouldn't this be `m_processName`?

csaavedra commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -173,10 +180,445 @@ enum TracePointCode {
+        }
+    }
+
+    static void Initialize(std::string processName)
```
This probably doesn't need to be a `std::string` and I would suggest to avoid passing by value to avoid copying.

csaavedra reviewed (commented):
Leaving a few minor comments about string usage, take with a grain of salt, I am not a reviewer :)

nikolaszimmermann commented on Source/JavaScriptCore/jit/JITPlan.cpp:
```diff
@@ -180,15 +180,15 @@ void JITPlan::compileInThread(JITWorklistThread* thread)
         stream.print(m_mode, " ", *m_codeBlock, " instructions size = ", m_codeBlock->instructionsSize());
         signpostMessage = stream.toCString();
-        WTFBeginSignpost(this, JSCJITCompiler, "%" PUBLIC_LOG_STRING, signpostMessage.data());
+        WTFBeginSignpost(this, JSCJITCompiler, "%" PUBLIC_LOG_STRING, signpostMessage.data() ?: "(nullptr)");
```
Haven't seen that shorthand for the ternary operator in WebKit so far, doesn't trigger a style violation, but still slightly awkward to my untrained eye :-)

Constellation replied:
I think this is GCC extension, and we should avoid it.

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
```
Not sure if this a good license template for Igalia-provided headers, there are others w/o that paragraph...

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
+public:
+
+    static void instantMark(const void*, const char *name, const char *description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
```
Slightly confusing to have a pointer as first argument, that's unused. Why not omit it, and adapt the call-sites?

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+
+    static void beginMark(const void* pointer, const char* name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
+    {
+        std::pair<const void*, const void*> key = std::make_pair(pointer, (const void*) name);
```
`auto key = std::make_pair<const void*, const void*>(pointer, name)` ? (eventually still needs a reinterpret_cast for the latter argument).

GeorgesStavracas replied:
`std::make_pair` is not a template function, but using `auto` and static casting sounds good.

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        vsnprintf(buffer, 1024, description, args);
+        va_end(args);
+
+        std::pair<int64_t, std::string> value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, buffer);
```
Ditto.

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+
+        std::pair<int64_t, std::string> value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, buffer);
+
+        auto& instance = getInstance();
```
Hmm, in general, I'd avoid manual locking/unlocking like the plague :-)
How about adding a e.g. `SysprofLocker` RAII class, that calls `instance.lock.lock()` in the CTOR, and unlock in the dtor?

carlosgcampos replied:
I don't think we need a new locker class, we can just use Locker, if this is an instance class we could just do:

```
Locker locker { m_lock };
m_ongoingMarks[key] = value;
```

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        }
+    }
+
+    static void Initialize(CString processName)
```
You explicitly want to copy here?

carlosgcampos replied:
```suggestion
    static void initialize(CString processName)
```

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    }
+
+private:
+    static const char* tracePointCodeName(TracePointCode code)
```
Use `ASCIILiteral` here.

nikolaszimmermann replied:
And I wish this was auto-generated (not asking you to do that now).

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        }
+    }
+
+    // This is actually <const void*, const char*>
```
That is better manifested in a using statement.
`using SomePair = std::pair<...>` to give it a descriptive name.

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    }
+
+    // This is actually <const void*, const char*>
+    std::map<std::pair<const void*, const void*>, std::pair<int64_t, std::string>> ongoingMarks;
```
`HashMap<SomePair, ASCIILiteral>` ? The traits are pre-existent.

carlosgcampos replied:
Just ASCIILIteral doesn't work, it needs another pair as value for the time value

nikolaszimmermann commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+
+    static SysprofAnnotator& getInstance()
+    {
+        static SysprofAnnotator instance;
```
Hmm, who is destructing the instance? The destructor runs a non-trivial amount of code. The static object destruction order is not guaranteed - this looks suspicious to me.

GeorgesStavracas replied:
The destructor is never called, this is a forever object. I've removed this code from the destructor because I reasoned that Sysprof is about reporting marks and timestamps of concrete, finished events, and these events in the HashMap are not finished.

mcatanzaro replied:
It actually is going to be destroyed in an exit handler, which is risky. Please include wtf/NeverDestroyed.h and then use `static NeverDestroyed<SysprofAnnotator> instance;` to avoid this.

nikolaszimmermann reviewed (changes requested):
Great job! Looking forward to have that in the tree..
Still a few things to cleanup IMHO.

carlosgcampos commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -328,6 +347,31 @@ enum WTFOSSignpostType {
+#elif USE(SYSPROF_CAPTURE)
+
+#define WTFEmitSignpost(pointer, name, ...) \
+    WTF::SysprofAnnotator::instantMark(pointer, _STRINGIFY(name), " " __VA_ARGS__)
```
Instead of using static members that get the instance, they could be instance memebers and the callers would do `WTF::SysprofAnnotator::singleton().foo()`, omitting the pointer param if we don't use it.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    {
+        va_list args;
+        va_start(args, description);
+        sysprof_collector_mark_vprintf(SYSPROF_CAPTURE_CURRENT_TIME, 0, getProcessName(), name, description, args);
```
If this is an instance method, we could just use m_processName.data(), we don't need a getter and it can be private.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+#include <wtf/text/CString.h>
+
+class SysprofAnnotator final {
+    WTF_MAKE_FAST_ALLOCATED;
```
We are not heap allocating it, no?

GeorgesStavracas replied:
No

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        va_list args;
+        va_start(args, description);
+        vsnprintf(buffer, 1024, description, args);
+        va_end(args);
```
If using `ASCIILiteral` isn't possible, then using `Vector<char>` we could avoid the memory copy, we copy into the vector buffer and then we move it to the pair

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+            value = WTFMove(instance.ongoingMarks[key]);
+            instance.ongoingMarks.erase(key);
+        }
+        instance.lock.unlock();
```
Using HashMaap this would be just `take()`

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        case WakeUpAndApplyDisplayListEnd:
+            return "WakeUpAndApplyDisplayList";
+
+
```
Extra line here

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        }
+    }
+
+    CString m_processName { "WebKit" };
```
Can this be used without initialization? It also calls the sysprof init function, so I guess it can't. It should be `ASCIILiteral` too instead of `CString`.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+            const void* pointer = std::get<0>(key);
+            const char* name = static_cast<const char*>(std::get<1>(key));
+            endMark(pointer, name, nullptr);
+        }
```
Have you checked the destructor is called? Maybe we can also add a finalize method to be called from web process platformFinalize.

GeorgesStavracas replied:
The destructor is never called, this is a forever object. I've removed this code from the destructor because I reasoned that Sysprof is about reporting marks and timestamps of concrete, finished events, and these events in the HashMap are not finished.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    std::map<std::pair<const void*, const void*>, std::pair<int64_t, std::string>> ongoingMarks;
+    Lock lock;
+
+    static SysprofAnnotator& getInstance()
```
In WebKit we use `singleton()`

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+
+    // This is actually <const void*, const char*>
+    std::map<std::pair<const void*, const void*>, std::pair<int64_t, std::string>> ongoingMarks;
+    Lock lock;
```
This should be m_lock. Also, I think it's easier to read if all instance members are at the end, after all the methods.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+    }
+
+    // This is actually <const void*, const char*>
+    std::map<std::pair<const void*, const void*>, std::pair<int64_t, std::string>> ongoingMarks;
```
This should be `m_ongoingMarks`. Also add `WTF_GUARDED_BY_LOCK(m_lock);`

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -211,6 +216,20 @@ void ProcessLauncher::launchProcess()
     g_subprocess_launcher_take_fd(launcher.get(), socketPair.client, socketPair.client);
 
+#if USE(SYSPROF_CAPTURE)
+    int sysprofFd = -1;
```
Use UnixFileDescritor

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -211,6 +216,20 @@ void ProcessLauncher::launchProcess()
+
+    if (sysprofFd >= 0) {
+        int copiedFd = dup(sysprofFd);
+        g_autofree char* fdStr = g_strdup_printf("%d", copiedFd);
```
`GUniquePtr<char>`

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+        return instance;
+    }
+
+    void instantMark(const char *name, const char *description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
```
```suggestion
    void instantMark(const char* name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
```

In WebKit the asterisks hang on the type, not the variable name. I'm surprised style checker doesn't catch this.

We're moving away from passing C-style strings as function parameters. You can use `std::span<char>` instead. It's more awkward and probably won't make this particular code any safer, but in general it's a good rule. Then you can access the underlying data of the span using e.g. `name.data()` for passing to the sysprof APIs.

GeorgesStavracas replied:
What are the performance implications of using `std::span` here? The Sysprof code is meant to run in very hot paths, the less constructs it uses the better, and I'm not sure `std::span` will help here.

mcatanzaro replied:
`std::span` is basically a pointer plus an integer (its size). You can copy it by value. It's as cheap as anything gets.

For unowned strings like this one, you basically want to use ASCIILiteral whenever possible (because it does bounds checking when accessed), and `std::span` otherwise. And you can't use ASCIILiteral here unless you're willing to touch the Cocoa implementation, so `std::span` is your best bet.

(`CString` is for when you need to own a simple C-style string, but that's not what you need here. And then `String` is WebKit's default full-featured string type, but it's not a C-style string.)

GeorgesStavracas replied:
I've changed the `name` variable to `std::span`. I couldn't change the others without breaking the `WTF_ATTRIBUTE_PRINTF` attribute.

I think the end result is a straight downgrade from the previous code, both in terms of legibility and in terms of performance, and I don't think security was increased here by any relevant metric. I'll leave it up to other maintainers to decide if they consider this an acceptable code that they wish to maintain.

mcatanzaro replied:
The security benefit is when standard library assertions are enabled (as they are in all Linux distros); it adds bounds checking to ensure WebKit will crash if you try to read an invalid index. I doubt there are any security bugs in your code here, but we're doing it for all of WebKit as a general hardening measure.

Anyway, I agree that keeping `WTF_ATTRIBUTE_PRINTF` is more important than getting rid of `const char *`, so this seems fine to me.

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    SysprofAnnotator() { };
+    ~SysprofAnnotator() { };
```
Slightly nicer:

```suggestion
    SysprofAnnotator()  = default;
```

I don't think you need to declare the destructor at all. (You do need to declare the constructor to ensure it is private.)

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+        case WebKit2Range:
+        case UIProcessRange:
+        case GPUProcessRange:
+        default:
```
I don't recommend using a default case here because it means nobody will add new enum values to the switch statement. If you remove the default case, then any new values will trip -Wswitch and the EWS should fail.

With the default case removed, you'll need to add a RELEASE_ASSERT_NOT_REACHED() after the switch statement to avoid tripping -Wreturn-type.

mcatanzaro commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -35,6 +35,14 @@
+#include <sysprof-capture.h>
+#include <wtf/Lock.h>
+#include <wtf/text/CString.h>
+#endif
```
This is a leftover from an earlier version of this commit, right? None of these are still needed here?

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+
+#pragma once
+
+#include <cstdio>
```
Is the cstdio include still needed? Probably not?

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+#include <wtf/Lock.h>
+#include <wtf/Vector.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/CString.h>
```
I think the CString include is also not needed.

mcatanzaro commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,437 @@
+    {
+        auto key = std::make_pair(pointer, static_cast<const void*>(name));
+
+        Vector<char, 1024> buffer;
```
Use std::array here. Vector is for when you need to change the size of the buffer, but here you're using a fixed size, so it's not needed.

I suppose this goes for TimestampAndString as well. Normally I would ask you to change TimestampAndString to use a normal String instead of this handrolled string, but I see Carlos requested it be done this way to avoid an extra copy. So I suppose he's going to prefer std::array. Whatever.

GeorgesStavracas replied:
Apparently std::array conflicts with HashMap, it complains about the size being too big and it's too inefficient to move when rehashing.

mcatanzaro replied:
Oh good point! That's reason enough to prefer Vector.

mcatanzaro reviewed (commented):
This looks good to me as well.

I tried to think of any way to avoid needing a singleton here, but couldn't, so I guess this is the best option.

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
```diff
@@ -211,6 +217,20 @@ void ProcessLauncher::launchProcess()
+    UnixFileDescriptor sysprofFd;
+
+    if (const char* sysprofFdStr = getenv("SYSPROF_CONTROL_FD"))
+        sysprofFd = UnixFileDescriptor(parseInteger<int>(StringView(std::span(sysprofFdStr, strlen(sysprofFdStr)))).value_or(-1), UnixFileDescriptor::Duplicate);
```
It will do a strlen for you when constructing the span:

```suggestion
        sysprofFd = UnixFileDescriptor(parseInteger<int>(StringView(std::span(sysprofFdStr))).value_or(-1), UnixFileDescriptor::Duplicate);
```

GeorgesStavracas replied:
```
Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:224:69: error: no viable constructor or deduction guide for deduction of template arguments of 'span'
  224 |         sysprofFd = UnixFileDescriptor(parseInteger<int>(StringView(std::span(sysprofFdStr))).value_or(-1), UnixFileDescriptor::Duplicate);
```

mcatanzaro replied:
OK, remove the `std::` and then it will work. It's actually coming from `WTF::span` in StringCommon.h.

Anyway, I'm going to click Approve now. Please make sure Carlos Garcia is OK with this version of the pull request before landing, though.

GeorgesStavracas replied:
It works! Thanks.

nikolaszimmermann reviewed (approved):
Looks great! Thank you for resolving all comments.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,436 @@
+        Locker locker { m_lock };
+        if (m_ongoingMarks.contains(key))
+            value = m_ongoingMarks.take(key);
+        locker.unlockEarly();
```
```suggestion
        std::optional<TimestampAndString> value;
        {
            Locker locker { m_lock };
            value = m_ongoingMarks.take(key);
        }
```

carlosgcampos replied:
We don't need contains + take, because take already returns the empty value (which for an optional should be nullopt) when the key is not present

GeorgesStavracas replied:
I'm not sure about removing `m_ongoingMarks.contains(key)`. When I remove it, `m_ongoingMarks.take(key)` returns a non-null `std::pair(0, "")`, which is different from `std::nullptr`. That's because the HashMap stores values of type `std::pair<int64_t, Vector<char>>` - not an optional - so the "empty" value is resolved to be `std::pair(0, "")`. In this function, I'm using an `std::optional` on top of the TimestampAndString type.

carlosgcampos replied:
Ah, optional is on top, still we can check if `value.first == 0` to know if the empty value has returned and avoid the double hash.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,436 @@
+    static SysprofAnnotator& singleton()
+    {
+        static LazyNeverDestroyed<SysprofAnnotator> instance;
+        return instance;
```
I wonder how this works, because LazyNeverDestroyed requires to explicitly call `construct()`, you should either use `std::once` here to construct the instance the first time `singleton()` is called or make the instance global and call `construct()` from `initialize()`

GeorgesStavracas replied:
I think this is fixed now

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,436 @@
+    static void initialize(ASCIILiteral processName)
+    {
+        singleton().m_processName = processName;
+        sysprof_collector_init();
```
So, if we take the approach of calling `construct()` from here, we can make process name a construct parameter and move `sysprof_collector_init()` to the constructor too.

carlosgcampos commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,436 @@
+    HashMap<RawPointerPair, TimestampAndString> m_ongoingMarks WTF_GUARDED_BY_LOCK(m_lock);
+};
+
+} // namespace WTF
```
For WTF types we normally add here using statement to avoid having to use `WTF::` 

```
using WTF::SysprofAnnotator;
```

GeorgesStavracas committed:
[WPE][GTK][WTF] Implement SystemTracing.h with Sysprof
https://bugs.webkit.org/show_bug.cgi?id=275314

Reviewed by Nikolas Zimmermann and Michael Catanzaro.

Implement the tracing macros provided by SystemTracing.h using the
Sysprof collector API.

Sysprof collector operates by passing a control fd as an environment
variable. This env var is set when running any executable with the
sysprof-cli utility, or through Sysprof's main interface.

The control fd needs to be propagated to child processes properly. For
the case of run-minibrowser script - a Python script that spawns the
actual MiniBrowser executable - and for the subprocess launcher. To do
that, duplicate the control fd and pass it as an environment variable
to subprocesses well. The Python script neatly supports that, but the
subprocess launcher needed some more manual mangling to get that done.

This initial work covers all pre-existing tracing groups implemented
on kdebug.

This patch does not introduce new tracing points.

There are limitations with the tracing macros though, for example,
they are most meaningful when there's a single WebKit view running.
Once more views are created, they all output into the same tracing
group. In the future, this can be fixed by using the subprocesses'
pids as part of the tracing name, but right now it's most useless
as the sandbox makes all subprocesses have pid 3.

Another limitation is that tracing only truly works when disabling
the sandboxing of subprocesses. Working around it might require some
reengineering of the subprocess launching mechanism, so this is left
out of this patch as well.

To run WebKitGTK with Sysprof tracing, run the following command in
the WebKit source directory:

```
$ sysprof-cli --force -e WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 -- Tools/Scripts/run-minibrowser --gtk
```

For WPE WebKit, run the following command:

```
$ sysprof-cli --force -e WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 -- Tools/Scripts/run-minibrowser --wpe
```

It is also possible to get GNOME Shell marks by passing `--gnome-shell`
to the sysprof-cli command.

Co-authored-by: Carlos Garcia Campos <cgarcia@igalia.com>

* Source/JavaScriptCore/jit/JITPlan.cpp:
(JSC::JITPlan::compileInThread):
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/SystemTracing.h:
(WTF::tracePoint):
* Source/WTF/wtf/glib/SysprofAnnotator.h: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setReadyState):
* Source/WebKit/GPUProcess/glib/GPUProcessMainGLib.cpp:
(WebKit::GPUProcessMain):
* Source/WebKit/UIProcess/API/glib/WebKitInitialize.cpp:
(WebKit::webkitInitialize):
* Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp:
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/WebProcess/gtk/WebProcessMainGtk.cpp:
* Source/WebKit/WebProcess/wpe/WebProcessMainWPE.cpp:
* Tools/Scripts/webkitpy/common/system/abstractexecutive.py:
(AbstractExecutive.run_command):
* Tools/Scripts/webkitpy/common/system/executive.py:
(Executive.run_command):
* Tools/Scripts/webkitpy/common/system/executive_mock.py:
(MockExecutive.run_command):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort.run_minibrowser):

Canonical link: https://commits.webkit.org/280690@main

kmiller68 commented on Source/WTF/wtf/glib/SysprofAnnotator.h:
```diff
@@ -0,0 +1,456 @@
+        auto key = std::make_pair(pointer, static_cast<const void*>(name.data()));
+
+        Vector<char, 1024> buffer;
+        va_list args;
+        va_start(args, description);
+        vsnprintf(buffer.data(), 1024, description, args);
+        va_end(args);
+
+        auto value = std::make_pair(SYSPROF_CAPTURE_CURRENT_TIME, WTFMove(buffer));
+
+        Locker locker { m_lock };
+        m_ongoingMarks.set(key, value);
```
Does this actually save any chars in the map? isn't the `size()` of `buffer` 0? So when it gets copied to the HashMap it will be empty, IIUC? I saw this when refactoring for https://github.com/WebKit/WebKit/pull/29744 but I just fixed the build error I didn't change the semantics here.

carlosgcampos replied:
I wondered the same when I reviewed this, and IIRC size it's not 0, because of the stack allocated size 1024.

mcatanzaro replied:
I'm not certain, but I'd guess capacity is 1024 but size is actually 0. Probably not good even if it works. I think we need to verify this.

carlosgcampos replied:
Yes, but we are not using `size()`, we pass 1024 directly to `vsnprintf()`. I agree we should probably resize the buffer to be sure. I guess we could just change the declaration to:

```
Vector<char, 1024> buffer(1024);
```

mcatanzaro replied:
@GeorgesStavracas note there is action required here

> Vector<char, 1024> buffer(1024);

I bet we don't need to specify the capacity at all if we construct it with a size.

cdumez replied:
> @GeorgesStavracas note there is action required here
> 
> > Vector<char, 1024> buffer(1024);
> 
> I bet we don't need to specify the capacity at all if we construct it with a size.

It does make a difference though, in one case, the bytes will be stack allocated, in the other heap allocated.

The second template parameter is the *inline* capacity, not the capacity. So the 1024 bytes would be allocated on the stack, as if you had a `char buffer[1024];` local variable.

mcatanzaro replied:
OK, that's pretty important, and I didn't know.

So this Vector is going to be moved into a hash table with a lifetime that extends beyond the current stack frame. I assume the data will be copied (and not left dangling) if using the inline capacity. So not using inline capacity seems better here, yes?

mcatanzaro replied:
https://github.com/WebKit/WebKit/pull/31816
